src/System.Private.ServiceModel/src/System/ServiceModel/Channels/TransportSecurityHelpers.cs

```diff
@@ -336,89 +337,61 @@ public static void RemoveIdentityMapping(Uri via, EndpointAddress target, bool v
             // support from HttpClient before any functionality can be added here. 
         }
 
-        private static Dictionary<HttpRequestMessage, string> s_serverCertMap = new Dictionary<HttpRequestMessage, string>();
-
-        public static void AddServerCertMapping(HttpRequestMessage request, EndpointAddress to)
+        public static void AddServerCertIdentityValidation(HttpClientHandler httpClientHandler, EndpointAddress to)
         {
-            Fx.Assert(request.RequestUri.Scheme == UriEx.UriSchemeHttps,
-                "Wrong URI scheme for AddServerCertMapping().");
             X509CertificateEndpointIdentity remoteCertificateIdentity = to.Identity as X509CertificateEndpointIdentity;
             if (remoteCertificateIdentity != null)
             {
                 // The following condition should have been validated when the channel was created.
                 Fx.Assert(remoteCertificateIdentity.Certificates.Count <= 1,
                     "HTTPS server certificate identity contains multiple certificates");
-                AddServerCertMapping(request, remoteCertificateIdentity.Certificates[0].Thumbprint);
-            }
-        }
+                var thumbprint = remoteCertificateIdentity.Certificates[0].Thumbprint;
+                bool identityValidator(HttpRequestMessage requestMessage, X509Certificate2 cert, X509Chain chain, SslPolicyErrors policyErrors)
+                {
+                    try
+                    {
+                        ValidateServerCertificate(cert, thumbprint);
```

Comment:
thumbprint isn't being checked for null anymore. I couldn't tell if it was checked before because we were retrieving something that may or may not have existed from s_serverCertMap... or because it might actually be null in some cases.

Reply:
I verified with @bartonjs that Thumprint can't ever be null, and in the process of the conversation came to the conclusion that this isn't the best implementation. I'm going to switch it to comparing the RawData (which also can't be null) of the certificate as it will be both faster and avoids hash collisions.

Reply:
@StephenMolloy. I've modified the approach can you re-review.